### ISSUE TITLE
Improve our project configurations

### ISSUE THREAD

**Issue by heytherewill** [toggl/mobileapp#4899]
## The problem
We have a single codebase that can have multiple possible outcomes. It can generate store artifacts or it can simply generate test results. This means we need to have a way to tell our build system what it needs to do depending on what outcome we expect.

## The way things are at the moment

We have multiple project configurations, basically one for each target of the apps (Release build for the Stores and Debug builds) and one for each type of test.

This approach has bitten us in the butt multiple times in the past because tooling doesn't play well with custom configurations:
- Visual Studio updates break the configurations
- Adding a new project generates all sorts of weirdness
- Opening the solution on Windows basically breaks everything

We also aren't really using the configurations inside the IDE to only build what we need; we use them as a way to tell the build script what it needs to do. This means that you still need to load iOS projects when working on Windows (even if a Mac is not connected). At first, this might seem like a good thing, but having the project loaded is not preventing people from committing changes that break only one or the other project's build.

## What I am proposing

We should ditch the custom configurations and use only the default ones (Debug and Release). We will then, instead, have multiple `.sln` files that relate to the possible outcomes of the app:

- `Toggl.sln`
  - The existing solution file
  - This will contain all of our projects and if people do not want 
- `Toggl.Droid.sln`
  - Contains all Unit tests, Android UI tests, the android project, and all shared code
  - Meant to be used for doing work on Android
- `Toggl.iOS.sln`
  - Contains all Unit test, iOS UI tests, the iOS project, the Siri project, and all shared code
  - Meant to be used for doing work on iOS
- `Toggl.Tests.sln`
  - Contains all unit tests, sync tests, integration tests and possibly other kinds of tests we might want to add in the future.
  - Meant to be used for the test flows of Bitrise (we can build the entire solution fast to see if tests are failing)
- `Toggl.Tests.UI.sln`
  - Contains Android and iOS UI Tests
  - These are meant to work with pre-built binaries, so I think including the actual projects is not needed
  - This solution is meant to be used by Bitrise to run our UI tests

Benefits of this approach:

- Fewer git problems when using Visual Studio on Windows
- No need to load projects you don't care about when working on app-specific features
- I checked and we won't have to change anything in our bitrise flows since most of it is handled by cake anyways; we just need to change the cake script

The only downside I can see from this approach is keeping the `.sln` projects up-to-date. However, it currently already is a problem to add a new project because configurations break, so this is not really an issue.

## TL;DR

I think we should ditch configurations and instead use specific `.sln` for generating our artifacts. If anyone has an objection to this, please voice your concerns

**Comment by amulware** [comment]
**About the stated problems**

I think all the problems mentioned comes down to people changing configurations without understanding them. If you leave them alone, they work just fine.

The main problem I see with them is that the way they are encoded in the solution file is a horrible pain. (And it doesn't help that IDEs like to edit them without asking - still, we shouldn't be committing changes without checking them.)

> having the project loaded is not preventing people from committing changes that break only one or the other project's build

This is merely stated as a fact. What's your reasoning?

**About the benefits listed**

> Fewer git problems when using Visual Studio on Windows

Which are exactly?

> No need to load projects you don't care about when working on app-specific features

You can already unload projects if you really want to, how does this help? Seems to only make it harder to switch what projects are loaded. What if I want to edit with all projects loaded but want to quickly build an iOS build, and then Android, or run a test, or.. Switching build configurations in the IDE is very easy.

**Other problems with the proposed solution**

I don't see how you are solving anything here since you still need plenty of build configurations.

Each app has debug, adhoc and release builds, we need a way of building everything, just to make sure it builds, we need to be able to build/run the ui tests of each platform separately. There's a whole bunch of different tests, do you just want to build all of them no matter what we're building? What about tools like the sync diagram generator?


---

Lastly, I think this is a plain wrong use of solutions and configurations. The way we are working right now is how configurations are intended to be used as far as I understand. Multiple solution files seems like a more clunky solution with plenty of problems of its own.

**Comment by heytherewill** [comment]
Closing as this is not priority atm
